### ISSUE TITLE
[QUESTION] Why helm filter by label after list?

### ISSUE THREAD

**Issue by dayeguilaiye** [helm/helm#12304]
When I reading helm source code, I found that when helm do `list` command with selector, helm will list all release and then filter them by labels:

https://github.com/helm/helm/blob/a3cbe35b5f081880e72b0d95a05c4b20d15bed90/pkg/action/list.go#L191-L195

 I think it could make helm client or who use helm SDK waste so much memory.

In client-go of k8s, I found that they do it in request params, not in client-side. This can provide users from previous problem.
https://github.com/kubernetes/client-go/blob/b8a03ab933ab332affa308e107cac58ccc6f40f4/kubernetes/typed/core/v1/secret.go#L84-L98

Is there any reason that helm do it after list, instead of add it to params when list `configmap` or `secret` ?
Should helm do it this way?

Output of `helm version`:
main

**Comment by dayeguilaiye** [comment]
@gjenkins8 , thanks for your reply!

> But I think if the equivalent behavior can be achieved more efficiently via server-side filter.

Yes, this is exactly what I mean. So I rise a issue to ask is there any reason helm do filtering on client-side.

If there is no specific reason, I'm happy to rise a PR to optimize this.

Should I wait for any contributor or maintainer assign this issue to me, or should I just rise a PR and wait for code review?

**Comment by gjenkins8** [comment]
> Should I wait for any contributor or maintainer assign this issue to me, or should I just rise a PR and wait for code review?

I can/have assigned yourself @dayeguilaiye. But generally, if someone would like to contribute a PR, we don't consider the assignment "binding" in any way. It might help avoid duplicate effort of course.

---
> If there is no specific reason, I'm happy to rise a PR to optimize this.

Just keep in mind, we need to maintain equivalent behavior. Helm is used in so many ways, even a small behavioral change will likely break many users.

**Comment by dayeguilaiye** [comment]
After studying the code of Helm, I realized that Helm implements its own logic for `labelSelector`, instead of querying through Kubernetes' `labelSelector`. 

https://github.com/helm/helm/blob/276121c8693b48978eae5c09602b1e74d9a2a7e6/pkg/action/list.go#L162-L195

As a result, when using the Driver (like `secret` or `configmap`) for list operations, the `labelSelector` is not directly passed. Instead, the filtering based on `labelSelector` happens only **after list all releases** in the namespace.

That could waste some memory when there are so many releases in a namespace while there are just few I want to query. And that't why I rise this issue.

I believe it's feasible to pass the `labelSelector` during the `list` operation(no matter driver is `configmap` or `secret`). However, this would require modifying the `Queryor` interface in `Driver` interface. 

https://github.com/helm/helm/blob/276121c8693b48978eae5c09602b1e74d9a2a7e6/pkg/storage/driver/driver.go#L89-L105

It might have to change `List` from `List(filter func(*rspb.Release) bool) ([]*rspb.Release, error)` to something like `List(filter func(*rspb.Release) bool, selector string) ([]*rspb.Release, error)`.

Considering that many users might have already implemented the `Driver` interface, I think such a modification might not be acceptable for Helm.

Nevertheless, I have managed to reduce the frequency of calls to this interface through other methods in our current setup. Therefore, the proposal is no longer necessary, and I will be closing this issue.
